In [ ]:
import pandas as pd
import numpy as np

In [ ]:
df = pd.read_csv('WA_Fn-UseC_-Telco-Customer-Churn.csv')

print("Dataset loaded successfully.")

Dataset loaded successfully.


In [3]:
# Print dataset shape
print(f"Dataset Shape: {df.shape}")
print(f"Number of rows: {df.shape[0]}")
print(f"Number of columns: {df.shape[1]}")

Dataset Shape: (7043, 21)
Number of rows: 7043
Number of columns: 21


In [4]:
# Print column names
print("Column Names:")
print("-" * 50)
for i, col in enumerate(df.columns, 1):
    print(f"{i:2d}. {col}")

Column Names:
--------------------------------------------------
 1. customerID
 2. gender
 3. SeniorCitizen
 4. Partner
 5. Dependents
 6. tenure
 7. PhoneService
 8. MultipleLines
 9. InternetService
10. OnlineSecurity
11. OnlineBackup
12. DeviceProtection
13. TechSupport
14. StreamingTV
15. StreamingMovies
16. Contract
17. PaperlessBilling
18. PaymentMethod
19. MonthlyCharges
20. TotalCharges
21. Churn


In [5]:
# Display first 5 rows
print("First 5 Rows:")
print("=" * 100)
df.head()

First 5 Rows:


,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
0,7590-VHVEG,Female,0,Yes,No,1,No,No phone service,DSL,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,29.85,29.85,No
1,5575-GNVDE,Male,0,No,No,34,Yes,No,DSL,Yes,...,Yes,No,No,No,One year,No,Mailed check,56.95,1889.5,No
2,3668-QPYBK,Male,0,No,No,2,Yes,No,DSL,Yes,...,No,No,No,No,Month-to-month,Yes,Mailed check,53.85,108.15,Yes
3,7795-CFOCW,Male,0,No,No,45,No,No phone service,DSL,Yes,...,Yes,Yes,No,No,One year,No,Bank transfer (automatic),42.30,1840.75,No
4,9237-HQITU,Female,0,No,No,2,Yes,No,Fiber optic,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,70.70,151.65,Yes


---

# Initial Dataset Orientation

## What type of company does this data come from?

This dataset comes from a **telecommunications (telco) company**. The filename itself includes "Telco-Customer-Churn", and based on the typical features found in such datasets, this would be a company providing:
- Phone services (landline and/or mobile)
- Internet services (DSL, fiber optic, cable)
- Potentially bundled services (phone + internet + TV)

The company likely operates on a subscription-based model where customers pay monthly for ongoing services.

---

## What unit of analysis does one row represent?

Each row represents a **single customer** (or customer account) at a specific point in time. 

This is customer-level data, where each record captures:
- Customer demographics (likely gender, age, dependents)
- Account information (tenure, contract type, payment method)
- Service subscriptions (what services they use)
- Billing information (charges, total spend)
- **Churn status** (whether the customer left or stayed)

---

## What is the *likely business problem* this dataset was created for?

The primary business problem is **customer churn prediction and retention strategy**.

**Why this matters to the business:**

1. **Churn is expensive**: Acquiring new customers costs 5-25x more than retaining existing ones in telco industries

2. **Predictive intervention**: The company wants to:
   - Identify which customers are at high risk of leaving
   - Understand *why* customers churn (which services, contracts, or behaviors correlate with churn)
   - Proactively intervene with targeted retention offers before customers cancel

3. **Customer Lifetime Value (LTV) optimization**: By understanding churn patterns, the company can:
   - Maximize revenue from each customer relationship
   - Allocate marketing/retention budgets efficiently
   - Design better service bundles and contract terms

**Expected use cases:**
- Build a predictive model to score customers by churn risk
- Segment customers for targeted retention campaigns
- Calculate Customer Lifetime Value based on churn probability
- Identify key drivers of churn to inform product/service improvements

---

---

# Column → Business Meaning Map

## Systematic Analysis of Each Column

Below is a detailed interpretation of every column in the dataset, categorized by business function.

In [7]:
# Display data types and sample values for context
print("Column Data Types and Sample Values:")
print("=" * 100)
print(f"\n{'Column Name':<20} {'Data Type':<15} {'Sample Values'}")
print("-" * 100)

for col in df.columns:
    dtype = str(df[col].dtype)
    # Get first non-null unique values (up to 3)
    sample_vals = df[col].dropna().unique()[:3]
    sample_str = ", ".join([str(v) for v in sample_vals])
    if len(sample_str) > 50:
        sample_str = sample_str[:47] + "..."
    print(f"{col:<20} {dtype:<15} {sample_str}")

Column Data Types and Sample Values:

Column Name          Data Type       Sample Values
----------------------------------------------------------------------------------------------------
customerID           str             7590-VHVEG, 5575-GNVDE, 3668-QPYBK
gender               str             Female, Male
SeniorCitizen        int64           0, 1
Partner              str             Yes, No
Dependents           str             No, Yes
tenure               int64           1, 34, 2
PhoneService         str             No, Yes
MultipleLines        str             No phone service, No, Yes
InternetService      str             DSL, Fiber optic, No
OnlineSecurity       str             No, Yes, No internet service
OnlineBackup         str             Yes, No, No internet service
DeviceProtection     str             No, Yes, No internet service
TechSupport          str             No, Yes, No internet service
StreamingTV          str             No, Yes, No internet service
StreamingMovie

## Detailed Column Interpretation

| # | Column Name | Real-Life Meaning | Category | Notes |
|---|-------------|-------------------|----------|-------|
| 1 | **customerID** | Unique identifier for each customer account | **IDENTIFIER** | Not useful for modeling; purely for tracking/linking records |
| 2 | **gender** | Customer's gender (Male/Female) | **Demographic** | Personal attribute; may correlate with service preferences |
| 3 | **SeniorCitizen** | Whether customer is 65+ years old (1=Yes, 0=No) | **Demographic** | Age-based segment; seniors may have different service needs/tech adoption |
| 4 | **Partner** | Whether customer has a partner/spouse (Yes/No) | **Demographic** | Household structure; may indicate stability/dual income |
| 5 | **Dependents** | Whether customer has dependents like children (Yes/No) | **Demographic** | Family obligations; affects budget priorities and service needs |
| 6 | **tenure** | Number of months the customer has been with the company | **Customer Behavior** | Loyalty indicator; longer tenure = more invested relationship |
| 7 | **PhoneService** | Whether customer subscribes to phone service (Yes/No) | **Customer Behavior** | Core service adoption; indicates what products customer uses |
| 8 | **MultipleLines** | Phone line configuration (Yes/No/No phone service) | **Customer Behavior** | Service intensity; multiple lines = higher engagement |
| 9 | **InternetService** | Type of internet service (DSL/Fiber optic/No) | **Customer Behavior** | Critical service choice; fiber = higher value customer |
| 10 | **OnlineSecurity** | Internet add-on: Online security service (Yes/No/No internet) | **Customer Behavior** | Premium feature adoption; indicates willingness to pay for extras |
| 11 | **OnlineBackup** | Internet add-on: Online backup service (Yes/No/No internet) | **Customer Behavior** | Premium feature adoption; data protection concern |
| 12 | **DeviceProtection** | Internet add-on: Device protection plan (Yes/No/No internet) | **Customer Behavior** | Premium feature adoption; insurance mindset |
| 13 | **TechSupport** | Internet add-on: Technical support service (Yes/No/No internet) | **Customer Behavior** | Premium feature adoption; may indicate less tech-savvy users |
| 14 | **StreamingTV** | Internet add-on: TV streaming service (Yes/No/No internet) | **Customer Behavior** | Entertainment service; competes with cable TV |
| 15 | **StreamingMovies** | Internet add-on: Movie streaming service (Yes/No/No internet) | **Customer Behavior** | Entertainment service; content consumption pattern |
| 16 | **Contract** | Contract term type (Month-to-month/One year/Two year) | **Contract/Commitment** | **CRITICAL**: Strongest churn predictor; commitment level |
| 17 | **PaperlessBilling** | Whether customer uses paperless billing (Yes/No) | **Customer Behavior** | Digital adoption; convenience vs. traditional preference |
| 18 | **PaymentMethod** | How customer pays (Electronic check/Mailed check/Bank transfer/Credit card) | **Customer Behavior** | Payment friction indicator; auto-pay = convenience, checks = manual effort |
| 19 | **MonthlyCharges** | Current monthly bill amount in dollars | **Pricing/Revenue** | Revenue per customer; higher charges may increase churn risk if value not perceived |
| 20 | **TotalCharges** | Total amount billed to customer over entire relationship | **Pricing/Revenue** | Lifetime revenue to date; function of tenure × average monthly charges |
| 21 | **Churn** | Whether customer left the company (Yes/No) | ** TARGET VARIABLE** | **THIS IS WHAT WE PREDICT**: Customer retention vs. attrition |

---

## Category Summary

Let me categorize all columns by their business function:

### 📋 **IDENTIFIERS** (Exclude from modeling)
- `customerID` — Unique account identifier

### 👤 **DEMOGRAPHICS** (Who is the customer?)
- `gender` — Male/Female
- `SeniorCitizen` — Age 65+ indicator (0/1)
- `Partner` — Has spouse/partner (Yes/No)
- `Dependents` — Has children/dependents (Yes/No)

### 🎯 **CUSTOMER BEHAVIOR** (How does the customer interact with the service?)
- `tenure` — Months as customer (loyalty/stickiness)
- `PhoneService` — Uses phone service (Yes/No)
- `MultipleLines` — Has multiple phone lines
- `InternetService` — Internet type (DSL/Fiber/No)
- `OnlineSecurity` — Add-on: Security service
- `OnlineBackup` — Add-on: Backup service
- `DeviceProtection` — Add-on: Device insurance
- `TechSupport` — Add-on: Technical support
- `StreamingTV` — Add-on: TV streaming
- `StreamingMovies` — Add-on: Movie streaming
- `PaperlessBilling` — Digital billing preference (Yes/No)
- `PaymentMethod` — Payment channel (Check/Bank/Card)

### 📜 **CONTRACT/COMMITMENT** (What's the agreement structure?)
- `Contract` — Contract duration (Month-to-month/1-year/2-year)

### 💰 **PRICING/REVENUE** (What's the financial relationship?)
- `MonthlyCharges` — Current monthly bill ($)
- `TotalCharges` — Cumulative charges to date ($)

### 🎯 **TARGET VARIABLE** (What we're predicting)
- `Churn` — Customer left the company (Yes/No)

---

## Key Identification Summary

### 🎯 **TARGET VARIABLE**
**`Churn`** — Binary outcome (Yes/No) indicating whether the customer terminated their service.

**Business interpretation**: This is the dependent variable we aim to predict. Understanding which customers will churn allows proactive retention efforts.

---

### 🚫 **IDENTIFIERS TO EXCLUDE FROM MODELING**
**`customerID`** — Unique string identifier for each customer account.

**Why exclude?** 
- No predictive power; it's an arbitrary assignment
- Including it would cause overfitting (model memorizes IDs instead of learning patterns)
- Only useful for tracking/joining data, not for statistical relationships

---

### 🔑 **KEY INSIGHT: Feature Distribution Across Categories**

| Category | Count | % of Features |
|----------|-------|---------------|
| Customer Behavior | 13 | 62% |
| Demographics | 4 | 19% |
| Pricing/Revenue | 2 | 9.5% |
| Contract/Commitment | 1 | 4.5% |
| Identifier | 1 | 4.5% |
| **Target Variable** | **1** | **--** |

**Observation**: The dataset is behavior-heavy (service adoption patterns), which suggests the company believes *how customers use services* is more predictive of churn than demographics alone.

In [8]:
# Programmatic verification of target variable and identifier
print("=" * 100)
print("VERIFICATION: Target Variable and Identifiers")
print("=" * 100)

# Check target variable
print(f"\n🎯 TARGET VARIABLE: 'Churn'")
print(f"   Data type: {df['Churn'].dtype}")
print(f"   Unique values: {df['Churn'].unique()}")
print(f"   Value counts:")
print(df['Churn'].value_counts())
print(f"   Churn rate: {(df['Churn'] == 'Yes').sum() / len(df) * 100:.2f}%")

# Check identifier
print(f"\n🔑 IDENTIFIER: 'customerID'")
print(f"   Data type: {df['customerID'].dtype}")
print(f"   Total rows: {len(df)}")
print(f"   Unique IDs: {df['customerID'].nunique()}")
print(f"   Is unique per row? {df['customerID'].nunique() == len(df)}")
print(f"   Sample IDs: {list(df['customerID'].head(3))}")

print("\n" + "=" * 100)

VERIFICATION: Target Variable and Identifiers

🎯 TARGET VARIABLE: 'Churn'
   Data type: str
   Unique values: <StringArray>
['No', 'Yes']
Length: 2, dtype: str
   Value counts:
Churn
No     5174
Yes    1869
Name: count, dtype: int64
   Churn rate: 26.54%

🔑 IDENTIFIER: 'customerID'
   Data type: str
   Total rows: 7043
   Unique IDs: 7043
   Is unique per row? True
   Sample IDs: ['7590-VHVEG', '5575-GNVDE', '3668-QPYBK']



---

## Business Logic Explanations

### Why These Categorizations Matter

#### **Demographics vs. Behavior**
- **Demographics** (gender, age, family) are *immutable* or slow-changing attributes
- **Behavior** (service adoption, usage patterns) reflects *choices* the customer makes
- For churn prediction: Behavior tends to be more actionable (we can influence it through offers/pricing)

#### **Contract/Commitment: The Most Critical Feature**
- `Contract` type is expected to be the **strongest churn predictor** because:
  - Month-to-month = zero switching cost → high churn risk
  - 1-year/2-year = penalty for leaving → low churn risk (until contract ends)
- This is a **structural** feature that directly impacts customer lock-in

#### **Revenue Features: Double-Edged Sword**
- `MonthlyCharges`: Higher bills *could* mean:
  - ✅ More value extracted (good customer)
  - ❌ Price sensitivity (churn risk if not getting perceived value)
- `TotalCharges`: Cumulative revenue ≈ tenure × avg monthly charges
  - Longer tenure customers have paid more, but also more invested in the relationship

#### **Add-On Services: Stickiness Indicators**
- Each "Yes" to OnlineSecurity, Backup, TechSupport, etc. represents:
  - **Revenue diversification** (beyond base internet/phone)
  - **Switching friction** (leaving means losing these features)
  - **Value perception** (willing to pay for extras = satisfied customer)
- Hypothesis: More add-ons → lower churn

#### **Payment Method: Proxy for Effort**
- Electronic check/Bank transfer/Credit card (auto-pay) = low friction
- Mailed check = manual, high effort → possible churn signal if burdensome
- This is a subtle behavioral indicator

---

**No transformations applied. This is pure interpretation of what exists in the raw data.**

---

# Data Quality Audit

## Phase 1: Diagnosis Only (No Fixes Applied)

This section systematically identifies data quality issues without making any corrections.

In [9]:
# Step 1: Inspect data types and missing values
print("=" * 100)
print("DATA TYPES AND MISSING VALUES INSPECTION")
print("=" * 100)

# Get data type information
dtypes_info = df.dtypes.to_frame(name='dtype')

# Count missing values (NaN, None, etc.)
missing_counts = df.isnull().sum().to_frame(name='missing_count')

# Calculate missing percentage
missing_pct = (df.isnull().sum() / len(df) * 100).to_frame(name='missing_pct')

# Combine into one dataframe
quality_df = dtypes_info.join(missing_counts).join(missing_pct)

print(f"\n{'Column':<25} {'Data Type':<15} {'Missing':<10} {'Missing %':<10}")
print("-" * 100)

for col in df.columns:
    dtype = str(quality_df.loc[col, 'dtype'])
    missing = int(quality_df.loc[col, 'missing_count'])
    missing_p = quality_df.loc[col, 'missing_pct']
    print(f"{col:<25} {dtype:<15} {missing:<10} {missing_p:<10.2f}%")

print("\n" + "=" * 100)

DATA TYPES AND MISSING VALUES INSPECTION

Column                    Data Type       Missing    Missing % 
----------------------------------------------------------------------------------------------------
customerID                str             0          0.00      %
gender                    str             0          0.00      %
SeniorCitizen             int64           0          0.00      %
Partner                   str             0          0.00      %
Dependents                str             0          0.00      %
tenure                    int64           0          0.00      %
PhoneService              str             0          0.00      %
MultipleLines             str             0          0.00      %
InternetService           str             0          0.00      %
OnlineSecurity            str             0          0.00      %
OnlineBackup              str             0          0.00      %
DeviceProtection          str             0          0.00      %
TechSupport  

In [10]:
# Step 2: Deep dive into suspicious columns
# Focus on TotalCharges, tenure, MonthlyCharges

print("=" * 100)
print("DETAILED INSPECTION: NUMERIC COLUMNS")
print("=" * 100)

suspicious_cols = ['tenure', 'MonthlyCharges', 'TotalCharges']

for col in suspicious_cols:
    print(f"\n🔍 Column: {col}")
    print("-" * 60)
    print(f"   Current dtype: {df[col].dtype}")
    print(f"   Missing (NaN): {df[col].isnull().sum()}")
    print(f"   Sample values: {list(df[col].head(5))}")
    print(f"   Unique count: {df[col].nunique()}")
    
    # Check if there are any non-numeric characters (for string columns)
    if df[col].dtype == 'object' or df[col].dtype == 'str':
        print(f"   ⚠️  WARNING: Expected numeric but found dtype={df[col].dtype}")
        # Show some non-empty unique values
        sample_unique = df[col].dropna().unique()[:10]
        print(f"   Sample unique values: {sample_unique}")

print("\n" + "=" * 100)

DETAILED INSPECTION: NUMERIC COLUMNS

🔍 Column: tenure
------------------------------------------------------------
   Current dtype: int64
   Missing (NaN): 0
   Sample values: [1, 34, 2, 45, 2]
   Unique count: 73

🔍 Column: MonthlyCharges
------------------------------------------------------------
   Current dtype: float64
   Missing (NaN): 0
   Sample values: [29.85, 56.95, 53.85, 42.3, 70.7]
   Unique count: 1585

🔍 Column: TotalCharges
------------------------------------------------------------
   Current dtype: str
   Missing (NaN): 0
   Sample values: ['29.85', '1889.5', '108.15', '1840.75', '151.65']
   Unique count: 6531
   ⚠️  WARNING: Expected numeric but found dtype=str
   Sample unique values: <StringArray>
[  '29.85',  '1889.5',  '108.15', '1840.75',  '151.65',   '820.5',  '1949.4',
   '301.9', '3046.05', '3487.95']
Length: 10, dtype: str



In [11]:
# Step 3: Check for hidden issues in TotalCharges
# String type with spaces or empty strings can masquerade as missing data

print("=" * 100)
print("SPECIAL CHECK: TotalCharges Hidden Formatting Issues")
print("=" * 100)

# Check if TotalCharges has empty strings or whitespace
print(f"\nChecking for empty/whitespace values in TotalCharges...")

# Count empty strings
empty_string_count = (df['TotalCharges'] == '').sum()
print(f"Empty strings (''): {empty_string_count}")

# Count whitespace-only strings
whitespace_count = df['TotalCharges'].apply(lambda x: isinstance(x, str) and x.strip() == '').sum()
print(f"Whitespace-only strings: {whitespace_count}")

# Show rows where TotalCharges might be problematic
print(f"\nRows where TotalCharges is empty or whitespace:")
problematic_rows = df[df['TotalCharges'].apply(lambda x: isinstance(x, str) and x.strip() == '')]
if len(problematic_rows) > 0:
    print(f"Found {len(problematic_rows)} rows with empty/whitespace TotalCharges")
    print("\nSample problematic rows:")
    print(problematic_rows[['customerID', 'tenure', 'MonthlyCharges', 'TotalCharges']].head(10))
else:
    print("No empty/whitespace values found")

# Try to convert to numeric to see what fails
print(f"\nAttempting numeric conversion test (errors='coerce')...")
try:
    numeric_test = pd.to_numeric(df['TotalCharges'], errors='coerce')
    conversion_failures = numeric_test.isnull().sum()
    original_nulls = df['TotalCharges'].isnull().sum()
    new_nulls = conversion_failures - original_nulls
    print(f"Original NaN values: {original_nulls}")
    print(f"NaN after conversion: {conversion_failures}")
    print(f"Values that failed conversion: {new_nulls}")
except Exception as e:
    print(f"Conversion test failed: {e}")

print("\n" + "=" * 100)

SPECIAL CHECK: TotalCharges Hidden Formatting Issues

Checking for empty/whitespace values in TotalCharges...
Empty strings (''): 0
Whitespace-only strings: 11

Rows where TotalCharges is empty or whitespace:
Found 11 rows with empty/whitespace TotalCharges

Sample problematic rows:
      customerID  tenure  MonthlyCharges TotalCharges
488   4472-LVYGI       0           52.55             
753   3115-CZMZD       0           20.25             
936   5709-LVOEQ       0           80.85             
1082  4367-NUYAO       0           25.75             
1340  1371-DWPAZ       0           56.05             
3331  7644-OMVMY       0           19.85             
3826  3213-VVOLG       0           25.35             
4380  2520-SGTTA       0           20.00             
5218  2923-ARZLG       0           19.70             
6670  4075-WKNIU       0           73.35             

Attempting numeric conversion test (errors='coerce')...
Original NaN values: 0
NaN after conversion: 11
Values that faile

In [12]:
# Step 4: Check expected vs actual data types
print("=" * 100)
print("EXPECTED vs ACTUAL DATA TYPES")
print("=" * 100)

# Define expected data types based on business logic
expected_types = {
    'customerID': 'string',
    'gender': 'categorical',
    'SeniorCitizen': 'numeric (binary)',
    'Partner': 'categorical',
    'Dependents': 'categorical',
    'tenure': 'numeric (int)',
    'PhoneService': 'categorical',
    'MultipleLines': 'categorical',
    'InternetService': 'categorical',
    'OnlineSecurity': 'categorical',
    'OnlineBackup': 'categorical',
    'DeviceProtection': 'categorical',
    'TechSupport': 'categorical',
    'StreamingTV': 'categorical',
    'StreamingMovies': 'categorical',
    'Contract': 'categorical',
    'PaperlessBilling': 'categorical',
    'PaymentMethod': 'categorical',
    'MonthlyCharges': 'numeric (float)',
    'TotalCharges': 'numeric (float)',
    'Churn': 'categorical (binary target)'
}

print(f"\n{'Column':<25} {'Expected Type':<25} {'Actual Type':<20} {'Match?':<10}")
print("-" * 100)

issues = []

for col in df.columns:
    expected = expected_types.get(col, 'unknown')
    actual = str(df[col].dtype)
    
    # Check if actual matches expected
    match = "✓"
    issue_note = ""
    
    if 'numeric' in expected.lower():
        if actual not in ['int64', 'float64', 'int32', 'float32']:
            match = "✗"
            issue_note = f"Should be numeric but is {actual}"
            issues.append((col, actual, expected, issue_note))
    
    print(f"{col:<25} {expected:<25} {actual:<20} {match:<10}")
    if issue_note:
        print(f"{'':>25} ⚠️  {issue_note}")

print("\n" + "=" * 100)

EXPECTED vs ACTUAL DATA TYPES

Column                    Expected Type             Actual Type          Match?    
----------------------------------------------------------------------------------------------------
customerID                string                    str                  ✓         
gender                    categorical               str                  ✓         
SeniorCitizen             numeric (binary)          int64                ✓         
Partner                   categorical               str                  ✓         
Dependents                categorical               str                  ✓         
tenure                    numeric (int)             int64                ✓         
PhoneService              categorical               str                  ✓         
MultipleLines             categorical               str                  ✓         
InternetService           categorical               str                  ✓         
OnlineSecurity            ca

In [13]:
# Step 5: Generate Data Quality Audit Table
print("=" * 100)
print("DATA QUALITY AUDIT SUMMARY TABLE")
print("=" * 100)

# Build audit table
audit_data = []

for col in df.columns:
    dtype = str(df[col].dtype)
    missing = df[col].isnull().sum()
    
    # Determine issue notes
    issue_notes = []
    
    # Check for type mismatches
    if col == 'TotalCharges' and dtype == 'str':
        issue_notes.append("String instead of numeric; contains empty strings")
    elif col in ['tenure', 'MonthlyCharges'] and dtype not in ['int64', 'float64']:
        issue_notes.append(f"Expected numeric, got {dtype}")
    
    # Check for missing values
    if missing > 0:
        issue_notes.append(f"{missing} missing values")
    
    # Check for empty strings in TotalCharges
    if col == 'TotalCharges':
        empty_count = df[col].apply(lambda x: isinstance(x, str) and x.strip() == '').sum()
        if empty_count > 0:
            issue_notes.append(f"{empty_count} empty/whitespace strings (hidden nulls)")
    
    issue_text = "; ".join(issue_notes) if issue_notes else "No issues detected"
    
    audit_data.append({
        'column': col,
        'dtype': dtype,
        'missing_count': missing,
        'issue_notes': issue_text
    })

# Create audit DataFrame
audit_df = pd.DataFrame(audit_data)

print("\n")
print(audit_df.to_string(index=False))
print("\n" + "=" * 100)

DATA QUALITY AUDIT SUMMARY TABLE


          column   dtype  missing_count                                                                                   issue_notes
      customerID     str              0                                                                            No issues detected
          gender     str              0                                                                            No issues detected
   SeniorCitizen   int64              0                                                                            No issues detected
         Partner     str              0                                                                            No issues detected
      Dependents     str              0                                                                            No issues detected
          tenure   int64              0                                                                            No issues detected
    PhoneService     str   

---

## Senior-Level Analysis: Data Quality Issues Explained

### 🔴 **Critical Issue #1: TotalCharges Data Type Mismatch**

**Observation:**
- **Expected:** `float64` (numeric)
- **Actual:** `str` (string/object)
- **Hidden problem:** Contains empty strings (`''`) instead of NaN for missing values

**Why This Exists:**
1. **CSV Import Quirk**: When the CSV was created, missing values in the TotalCharges column were likely exported as empty strings (`''`) instead of proper null markers
2. **Pandas Default Behavior**: When pandas encounters mixed content (numbers + empty strings), it defaults to treating the entire column as string/object type to preserve all data
3. **Common Data Export Flaw**: The source system (likely a database export or Excel export) didn't properly handle null values

**Why This Matters for Modeling:**
1. **Mathematical Operations Fail**: Cannot calculate statistics (mean, std, correlation) on string data
2. **Model Incompatibility**: Machine learning algorithms require numeric input; strings will cause errors or force the column to be dropped
3. **Hidden Missing Data**: Empty strings are *semantically* missing but *technically* not null, so `df.isnull()` won't catch them
4. **Business Logic Impact**: `TotalCharges = tenure × average MonthlyCharges` — we need this as numeric to validate data consistency

**Expected Behavior After Fix:**
- Convert to `float64`
- Empty strings → `NaN`
- Then handle missing values appropriately (likely impute or investigate why they're missing)

---

### 🟡 **Issue #2: Empty Strings in TotalCharges (Hidden Nulls)**

**Observation:**
- Some rows have `TotalCharges = ''` (empty string) instead of a numeric value or explicit `NaN`
- These rows are *not* counted as missing by standard pandas null checks

**Why This Exists:**
1. **New Customers Hypothesis**: Customers with `tenure = 0` or very short tenure may not have been billed yet
2. **Data Collection Timing**: Snapshot was taken before first billing cycle completed
3. **Data Export Bug**: System malfunction or incomplete record

**Why This Matters for Modeling:**
1. **Invisible Missing Data**: Standard missing value analysis (`df.isnull()`) won't detect these
2. **Biased Statistics**: If we convert to numeric without handling empties, we'll introduce NaNs that skew our missing data percentage
3. **Pattern Detection**: These rows might represent a specific customer segment (e.g., very new customers) that we need to handle specially

**Investigation Needed:**
- Check if all empty TotalCharges rows have `tenure = 0` or very low tenure
- Determine if these should be imputed as `0.0` or based on `MonthlyCharges × tenure`

---

### ✅ **Non-Issue #3: tenure and MonthlyCharges**

**Observation:**
- `tenure`: Correctly stored as `int64` ✓
- `MonthlyCharges`: Correctly stored as `float64` ✓
- No missing values (both show 0 nulls) ✓

**Why This is Good:**
- These columns are ready for mathematical operations
- No type conversion needed
- No missing data to impute

---

### 🟢 **Expected Behavior: Categorical Variables as Strings**

**Observation:**
- Most columns (`gender`, `Partner`, `Contract`, `Churn`, etc.) are stored as `str` (object dtype)

**Why This is Acceptable (For Now):**
1. **Pandas Default**: String/object is the default for categorical text data
2. **Not an Issue Yet**: We haven't reached the modeling stage where we need numeric encoding
3. **Readable**: Keeps data interpretable during EDA phase

**Future Consideration:**
- During preprocessing, these will need to be:
  - Label encoded (for tree-based models)
  - One-hot encoded (for linear models)
  - Or kept as pandas `category` dtype for efficiency

**Not a quality issue — just a note for Phase 2 (preprocessing).**

---

### 📊 **Missing Values: Overall Assessment**

**Observation:**
- Most columns show `0` missing values
- Only `TotalCharges` has hidden missing data (empty strings)

**Why This is Unusual:**
- Real-world datasets typically have more missing data
- This *could* indicate:
  - High-quality data collection
  - Pre-cleaned dataset (someone already handled nulls)
  - Synthetic/curated data for educational purposes

**Why This Matters:**
- If this is a curated dataset, it might not reflect real-world messiness
- In production, we'd expect missing values in demographics, service features, etc.
- Good news: Less imputation work needed

---

### 🎯 **Summary of Findings**

| Severity | Issue | Columns Affected | Action Required |
|----------|-------|------------------|-----------------|
| **CRITICAL** | String type instead of numeric | `TotalCharges` | Convert to float64; handle empty strings |
| **HIGH** | Hidden null values (empty strings) | `TotalCharges` | Investigate pattern; convert to NaN; decide imputation strategy |
| **LOW** | Categorical stored as object | Most columns | Future: Encode for modeling (not urgent) |
| **NONE** | Missing values | All columns (except TotalCharges) | No immediate action |

---

**Next Steps (NOT executed in this audit):**
1. Investigate *why* TotalCharges has empty strings (tenure correlation?)
2. Convert TotalCharges to numeric with proper null handling
3. Decide imputation strategy (0.0 vs. calculated vs. drop rows)
4. Validate TotalCharges ≈ tenure × MonthlyCharges for consistency checks

In [14]:
# Final diagnostic: Show the pattern of problematic rows
print("=" * 100)
print("PATTERN ANALYSIS: Why do 11 rows have empty TotalCharges?")
print("=" * 100)

problematic_rows = df[df['TotalCharges'].apply(lambda x: isinstance(x, str) and x.strip() == '')]

print(f"\nAll {len(problematic_rows)} rows with empty TotalCharges:\n")
print(problematic_rows[['customerID', 'tenure', 'MonthlyCharges', 'TotalCharges', 'Contract', 'Churn']].to_string(index=False))

print(f"\n\n{'Statistic':<30} {'Value'}")
print("-" * 60)
print(f"{'All have tenure = 0?':<30} {(problematic_rows['tenure'] == 0).all()}")
print(f"{'Mean tenure:':<30} {problematic_rows['tenure'].mean():.2f} months")
print(f"{'Mean MonthlyCharges:':<30} ${problematic_rows['MonthlyCharges'].mean():.2f}")
print(f"{'Churn rate in this group:':<30} {(problematic_rows['Churn'] == 'Yes').sum() / len(problematic_rows) * 100:.1f}%")
print(f"{'Overall churn rate:':<30} {(df['Churn'] == 'Yes').sum() / len(df) * 100:.1f}%")

print("\n" + "=" * 100)
print("CONCLUSION: All 11 rows with empty TotalCharges have tenure = 0 (brand new customers)")
print("These customers have not completed their first billing cycle yet.")
print("=" * 100)

PATTERN ANALYSIS: Why do 11 rows have empty TotalCharges?

All 11 rows with empty TotalCharges:

customerID  tenure  MonthlyCharges TotalCharges Contract Churn
4472-LVYGI       0           52.55              Two year    No
3115-CZMZD       0           20.25              Two year    No
5709-LVOEQ       0           80.85              Two year    No
4367-NUYAO       0           25.75              Two year    No
1371-DWPAZ       0           56.05              Two year    No
7644-OMVMY       0           19.85              Two year    No
3213-VVOLG       0           25.35              Two year    No
2520-SGTTA       0           20.00              Two year    No
2923-ARZLG       0           19.70              One year    No
4075-WKNIU       0           73.35              Two year    No
2775-SEFEE       0           61.90              Two year    No


Statistic                      Value
------------------------------------------------------------
All have tenure = 0?           True
Mean tenure

---

## 📋 Final Data Quality Audit Report

### Executive Summary

**Dataset Quality:** Generally high, with one critical issue requiring attention.

### Issues Identified (Ranked by Severity)

#### 🔴 **CRITICAL - Issue #1**
**Column:** `TotalCharges`  
**Problem:** Stored as `str` instead of `float64`; contains 11 empty/whitespace strings  
**Root Cause:** All 11 rows are brand new customers (`tenure = 0`) who haven't been billed yet  
**Impact:** Blocks all numeric operations; prevents correlation analysis, feature engineering, and modeling  
**Fix Strategy (Phase 2):** Convert to numeric; empty strings → `NaN` or `0.0`  

#### ✅ **NO ISSUES - Numeric Columns**
**Columns:** `tenure`, `MonthlyCharges`, `SeniorCitizen`  
**Status:** Correct data types (`int64`, `float64`); zero missing values  

#### ✅ **EXPECTED BEHAVIOR - Categorical Columns**
**Columns:** All other columns (gender, Contract, Churn, etc.)  
**Status:** Stored as `str`; appropriate for current phase; will require encoding in Phase 2  

### Key Statistics
- **Total Rows:** 7,043
- **Total Columns:** 21
- **Columns with True Missing Values:** 0
- **Columns with Hidden Missing Values:** 1 (`TotalCharges` with 11 empty strings)
- **Data Quality Score:** 95% (20/21 columns ready to use)

### Interesting Findings
1. **New Customer Segment:** 11 customers (0.16%) have `tenure = 0` — all signed up recently
2. **Zero Churn in New Customers:** None of the 11 new customers have churned (expected — they just joined)
3. **Commitment Pattern:** 10/11 new customers chose 2-year contracts (90.9%) — suggests aggressive retention strategy for new signups

### Next Phase Requirements
Before modeling, we must:
1. ✅ **Verified:** Understand column meanings (COMPLETE)
2. ✅ **Verified:** Identify data quality issues (COMPLETE)  
3. ⏳ **Pending:** Fix TotalCharges data type and handle nulls (Phase 2)
4. ⏳ **Pending:** Encode categorical variables (Phase 2)
5. ⏳ **Pending:** Feature engineering and scaling (Phase 2)

**Status:** Phase 1 (Data Ingestion & Audit) COMPLETE ✅  
**Ready for:** Phase 2 (Data Cleaning & Preprocessing)